### PR TITLE
HPCC-19489 STD.File.RenameLogicalFile should support OVERWRITE

### PR DESCRIPTION
**Pull request by AttilaVamos**
Add allowOverwrite parameter to STD.File.RenameLogicalFile

Add code if overwrite is allowed to:
- check source and target file existence
- remove existing target file

Extend fileservice.ecl to test allowOverwrite option

Signed-off-by: Attila Vamos <attila.vamos@gmail.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [x] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [x] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [x] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Testing:

Changes are tested manually.

### CODE REVIEW

**Comment by AttilaVamos** [comment]
@jakesmith please review

File: plugins/fileservices/fileservices.cpp [hunk 1]
```diff
@@ -523,31 +523,63 @@ FILESERVICES_API void FILESERVICES_CALL fsSetReadOnly(ICodeContext *ctx, const c
     throw error.getClear();
 }
 
-
-FILESERVICES_API void FILESERVICES_CALL fsRenameLogicalFile(ICodeContext *ctx, const char *oldname, const char *newname)
+FILESERVICES_API void FILESERVICES_CALL implementRenameLogicalFile(ICodeContext *ctx, const char *oldname, const char *newname, const bool overwrite)
 {
     StringBuffer lfn, nlfn;
```
Comment (jakesmith) [inline plugins/fileservices/fileservices.cpp]: pedantic/not new: but would be slightly clearer if lfn was olfn, or these were oldLfn/newLfn or similar.

File: plugins/fileservices/fileservices.cpp [hunk 2]
```diff
@@ -523,31 +523,63 @@ FILESERVICES_API void FILESERVICES_CALL fsSetReadOnly(ICodeContext *ctx, const c
...
     constructLogicalName(ctx, newname, nlfn);
 
     IDistributedFileTransaction *transaction = ctx->querySuperFileTransaction();
     Linked<IUserDescriptor> udesc = ctx->queryUserDescriptor();
+
+    IDistributedFileDirectory &distributedDirectory = queryDistributedFileDirectory();
+    if (overwrite)
+    {
+        if ( distributedDirectory.exists(nlfn.str(), ctx->queryUserDescriptor(), false, false) &&
+             distributedDirectory.exists(lfn.str(), ctx->queryUserDescriptor(), false, false))
```
Comment (jakesmith) [inline plugins/fileservices/fileservices.cpp]: If old lfn doesn't exist, there's no point in continuing at all, with or without overwrite. (renamePhysical will fail).
So would be clearer if checked if old existed and error out before anything else.

File: plugins/fileservices/fileservices.cpp [hunk 3]
```diff
@@ -523,31 +523,63 @@ FILESERVICES_API void FILESERVICES_CALL fsSetReadOnly(ICodeContext *ctx, const c
...
+            {
+                StringBuffer s("DeleteLogicalFile ('");
+                s.append(nlfn);
+                if (transaction->active())
+                    s.append("') added to transaction");
+                else
+                    s.append("') done");
+                WUmessage(ctx, SeverityInformation, NULL, s.str());
+                AuditMessage(ctx, "DeleteLogicalFile", nlfn.str());
+            }
```
Comment (jakesmith) [inline plugins/fileservices/fileservices.cpp]: lines 541 to 552 are same as code in fsDeleteLogicalFile.
should be commoned up, i.e. both call common function, or possibly for this just to calll fsDeleteLogicalFile directly.

File: testing/regress/ecl/fileservice.ecl [hunk 4]
```diff
@@ -33,5 +33,10 @@ SEQUENTIAL(
   File.RenameLogicalFile(prefix + 'afterrename1.d00', prefix + 'scope1::scope2::afterrename2.d00'),
   File.RenameLogicalFile(prefix + 'scope1::scope2::afterrename2.d00', prefix + 'scope1::afterrename3.d00'),
   OUTPUT(DATASET(prefix + 'scope1::afterrename3.d00', rec, FLAT)),
-  File.DeleteLogicalFile(prefix + 'scope1::afterrename3.d00')
+  File.DeleteLogicalFile(prefix + 'scope1::afterrename3.d00'),
+  OUTPUT(ds, , prefix + 'renametest.d00', OVERWRITE),
+  OUTPUT(ds, , prefix + 'afterrename1.d00', OVERWRITE),
+  // Remane with overwrite allowed
+  File.RenameLogicalFile(prefix + 'renametest.d00', prefix + 'afterrename1.d00', true),
```
Comment (jakesmith) [inline testing/regress/ecl/fileservice.ecl]: I think there may be some untested problems.
1) Rename external or foreign with overwrite.
Both are disallowed and will error on rename, but the delete will have already happened.. 
Foreign delete should fail, so may be ok.
External delete .. I'm not sure, worth testing.
i.e. need to make sure it doesn't succeed to delete - but then fail on rename.
2) Rename in a transaction.
Actually fileservices.ecl doesn't have any transaction tests at the moment.
I think there may be problems with rename in particular, related to the way delete happens.
See addDelayedDelete/CDelayedDelete in dadfs.cpp

File: testing/regress/ecl/fileservice.ecl [hunk 5]
```diff
@@ -33,5 +33,10 @@ SEQUENTIAL(
   File.RenameLogicalFile(prefix + 'afterrename1.d00', prefix + 'scope1::scope2::afterrename2.d00'),
   File.RenameLogicalFile(prefix + 'scope1::scope2::afterrename2.d00', prefix + 'scope1::afterrename3.d00'),
   OUTPUT(DATASET(prefix + 'scope1::afterrename3.d00', rec, FLAT)),
-  File.DeleteLogicalFile(prefix + 'scope1::afterrename3.d00')
+  File.DeleteLogicalFile(prefix + 'scope1::afterrename3.d00'),
+  OUTPUT(ds, , prefix + 'renametest.d00', OVERWRITE),
+  OUTPUT(ds, , prefix + 'afterrename1.d00', OVERWRITE),
+  // Remane with overwrite allowed
```
Comment (jakesmith) [inline testing/regress/ecl/fileservice.ecl]: typo: Rename

File: testing/regress/ecl/fileservice.ecl [hunk 6]
```diff
@@ -33,5 +33,10 @@ SEQUENTIAL(
...
+  File.RenameLogicalFile(prefix + 'renametest.d00', prefix + 'afterrename1.d00', true),
+  File.DeleteLogicalFile(prefix + 'afterrename1.d00'),
```
Comment (jakesmith) [inline testing/regress/ecl/fileservice.ecl]: Would be good to read the renamed file after the test to verify before deleting it.

**Review by jakesmith — CHANGES_REQUESTED** [review]

**Comment by jakesmith** [comment]
@AttilaVamos - added some comments.

File: plugins/fileservices/fileservices.cpp (hunk 1 above)
Comment (AttilaVamos) [reply plugins/fileservices/fileservices.cpp]: Changed

File: plugins/fileservices/fileservices.cpp (hunk 2 above)
Comment (AttilaVamos) [reply plugins/fileservices/fileservices.cpp]: Changed

File: plugins/fileservices/fileservices.cpp (hunk 3 above)
Comment (AttilaVamos) [reply plugins/fileservices/fileservices.cpp]: Changed

File: testing/regress/ecl/fileservice.ecl (hunk 5 above)
Comment (AttilaVamos) [reply testing/regress/ecl/fileservice.ecl]: Changed

File: testing/regress/ecl/fileservice.ecl (hunk 4 above)
Comment (AttilaVamos) [reply testing/regress/ecl/fileservice.ecl]: As we discussed, I will test this delete, rename in transaction.

File: testing/regress/ecl/fileservice.ecl (hunk 6 above)
Comment (AttilaVamos) [reply testing/regress/ecl/fileservice.ecl]: Added

**Comment by AttilaVamos** [comment]
@jakesmith updated, please re-review.

**Comment by jakesmith** [comment]
@AttilaVamos - looks good, but 1 question.

Did the "As we discussed, I will test this delete, rename in transaction."  happen, is there a separate JIRA?

**Comment by richardkchapman** [comment]
@AttilaVamos Please squash

**Comment by AttilaVamos** [comment]
@jakesmith I will open a JIRA or perhaps two to cover rename logical file in transaction, for foreign and external file.

**Comment by jakesmith** [comment]
@AttilaVamos - please squash ready for merge and reference the new JIRA's in this issues JIRA.

**Pull request merged**